src/Dialect/ONNX/Transforms/DQBinaryQOpt.cpp

```diff
@@ -0,0 +1,573 @@
+//===- foldDqBinaryQPattern.cpp - Remove DQ-Binary-Q chains -----*- C++ -*-===//
+//
+// (c) Copyright 2022 - 2025 Advanced Micro Devices, Inc. All Rights Reserved.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/WalkPatternRewriteDriver.h"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Dialect/ONNX/ONNXOps/OpHelper.hpp"
+#include "src/Pass/Passes.hpp"
+#include "llvm/ADT/STLExtras.h"
+#include <cmath> // For std::llround
+#include <optional>
+#include <variant>
+
+using namespace mlir;
+using namespace onnx_mlir;
+
+namespace {
+
+static ElementsAttr getElementAttributeFromConstant(Value val) {
+  if (!val)
+    return nullptr;
+  if (auto constOp = val.getDefiningOp<ONNXConstantOp>())
+    return mlir::dyn_cast<ElementsAttr>(constOp.getValueAttr());
+  return nullptr;
+}
+
+// Equivalent to Python's NoMatch exception: here Nullopt indicates failure.
+template <typename T>
+std::optional<T> get_scalar_tensor_value(ONNXConstantOp constOp) {
+  auto elementsAttr = dyn_cast_or_null<ElementsAttr>(constOp.getValueAttr());
+  if (!elementsAttr)
+    return std::nullopt;
+
+  Type elementType = elementsAttr.getElementType();
+
+  // Fast path: splat
+  if (elementsAttr.isSplat()) {
+    if (elementType.isa<FloatType>()) {
+      if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
+        APFloat splatValue = elementsAttr.getSplatValue<APFloat>();
+        return static_cast<T>(splatValue.convertToDouble());
+      }
+    }
+    if (auto intType = elementType.dyn_cast<IntegerType>()) {
```

Comment:
the other style will be likely deprecated later.
```suggestion
    if (auto intType = dyn_cast<IntegerType>(elementType)) {
```